packages/core/snapshot/snapshot.go

```diff
@@ -1,82 +1,150 @@
-			spentOutput.SetID(spentOutput.M.OutputID)
-			spentOutput.Output().SetID(spentOutput.M.OutputID)
-		}
-		for _, createdOutput := range epochdiff.Created() {
-			createdOutput.SetID(createdOutput.M.OutputID)
-			createdOutput.Output().SetID(createdOutput.M.OutputID)
-		}
+// LoadStreamableSnapshot loads a snapshot file from the given file path. Contents in a snapshot file
+// will not be written to a snapshot struct in case blowing up the memory, they should be proccessed in
+// consumer functions. To construct a snapshot struct from a file, use FromBytes([]byte).
+func LoadStreamableSnapshot(filePath string,
+	outputWithMetadataConsumer OutputWithMetadataConsumerFunc,
+	epochDiffsConsumer EpochDiffsConsumerFunc,
+	notarizationConsumer NotarizationConsumerFunc) (err error) {
+
+	f, err := os.Open(filePath)
+	defer f.Close()
+	if err != nil {
+		return fmt.Errorf("fail to open the snapshot file")
 	}
 
-	return nil
+	err = StreamSnapshotDataFrom(f, outputWithMetadataConsumer, epochDiffsConsumer, notarizationConsumer)
+
+	return
 }
 
+// Write a snapshot struct in to a file with the given file name.
 func (s *Snapshot) WriteFile(fileName string) (err error) {
-	if err = os.WriteFile(fileName, s.Bytes(), 0o644); err != nil {
+	data, err := s.Bytes()
+	if err != nil {
+		return err
+	}
+
+	if err = os.WriteFile(fileName, data, 0o644); err != nil {
 		return errors.Errorf("failed to write snapshot file %s: %w", fileName, err)
 	}
 
 	return nil
 }
 
 // Bytes returns a serialized version of the Snapshot.
-func (s *Snapshot) Bytes() (serialized []byte) {
-	return marshalutil.New().
-		WriteBytes(lo.PanicOnErr(serix.DefaultAPI.Encode(context.Background(), s.LedgerSnapshot))).
-		Bytes()
+func (s *Snapshot) Bytes() (serialized []byte, err error) {
+	marshaler := marshalutil.New()
+	header := s.LedgerSnapshot.Header
+
+	marshaler.
+		WriteUint64(header.OutputWithMetadataCount).
+		WriteInt64(int64(header.FullEpochIndex)).
+		WriteInt64(int64(header.DiffEpochIndex))
+
+	data, err := header.LatestECRecord.Bytes()
+	if err != nil {
+		return nil, err
+	}
+	marshaler.WriteBytes(data).WriteBytes(delimiter)
+
+	// write outputWithMetadata
+	typeSet := new(serix.TypeSettings)
+	data, err = serix.DefaultAPI.Encode(context.Background(), s.LedgerSnapshot.OutputsWithMetadata, serix.WithTypeSettings(typeSet.WithLengthPrefixType(serix.LengthPrefixTypeAsUint32)), serix.WithValidation())
+	if err != nil {
+		return nil, err
+	}
+	marshaler.WriteBytes(data).WriteBytes(delimiter)
+
+	// write epochDiffs
+	data, err = serix.DefaultAPI.Encode(context.Background(), s.LedgerSnapshot.EpochDiffs, serix.WithTypeSettings(typeSet.WithLengthPrefixType(serix.LengthPrefixTypeAsUint32)), serix.WithValidation())
+	if err != nil {
+		return nil, err
+	}
+	marshaler.WriteBytes(data).WriteBytes(delimiter)
+
+	return marshaler.Bytes(), nil
+}
+
+// FromBytes returns a Snapshot struct.
+func (s *Snapshot) FromBytes(data []byte) (err error) {
```

Comment:
This method assumes a non-stremeable snapshot (as it loads the entirety of s.LedgerSnapshot.OutputsWithMetadata at once). Why do we need it?

Reply:
We use it in `tools/genesis-snapshot` to test if the snapshot can be deserialized correctly.

Reply:
Can we move it there then? It is better to remove all "non-streamable" snapshot reading / writing out of the main codebase.